run_tests.py.in

```diff
@@ -0,0 +1,693 @@
+    """Call when new processes start.
+
+    This function is used as an initializer on a per-process basis due
+    to 'spawn' process strategy (at least on Windows and macOs).
+    """
+    _ctx.base_path = abshere(base_path)
+    _ctx.skeleton = skeleton
+    _ctx.keep_temp_files = keep_temp_files
+
+    # Setup logger for use within multiprocessing pool
+    _ctx.reporter = create_reporter(base_path)
+
+    # Set include paths, relative to build directory
+    cwd = os.getcwd()
+    os.chdir(_ctx.base_path)
+    paths = Path('.').rglob('*')
+    _ctx.incpaths = sum([['-I', str(d)] for d in paths if os.path.isdir(d)], [])
+    os.chdir(cwd)
+
+    if os.path.isabs(TOP_SRCDIR):
+        _ctx.incpaths += ['-I', f'{TOP_SRCDIR}/include']
+    else:
+        _ctx.incpaths += ['-I', f'../{TOP_SRCDIR}/include']
+
+    # Set tools paths
+
+    _ctx.diff = which('diff')[0]
+    if valgrind:
+        _ctx.valgrind = which('valgrind')[0]
+        _ctx.valgrind_opts = [
+            '-q',
+            '--track-origins=yes',
+            '--num-callers=50',
+            '--leak-check=full',
+            '--show-reachable=yes',
+            '--malloc-fill=0xa1',
+            '--free-fill=0xa1',
+        ]
+    else:
+        _ctx.valgrind_opts = []
+
+    if wine:
+        _ctx.wine = which('wine')[0]
+        _ctx.re2c += '.exe'
+        # Ignore whitespace at the end of line
+        _ctx.diff += ' -b'
+    elif os.name == 'nt':
+        # This should find 're2c.exe'
+        _ctx.re2c += which('re2c', path=os.getcwd())[0]
+        _ctx.diff += ' -b'
+
+
+def environment_check(valgrind=False, wine=False):
+    """Check environment and requirements."""
+    if not which('diff'):
+        raise RuntimeError("The 'diff' utility is missing from PATH")
+
+    if valgrind and not which('valgrind'):
+        raise RuntimeError("The 'valgrind' tool is missing from the PATH")
+
+    if wine and not which('wine'):
+        raise RuntimeError("The 'wine' layer is missing from the PATH")
+
+
+def main():
+    """The main function to run tests from the command line."""
+    args = parse_cli_arguments()
+
+    # Check environment
+    environment_check(args.valgrind, args.wine)
+
+    jobs = abs(args.jobs or 1)
+    print(f'Running in {jobs} job(s)')
+
+    if not os.access(_ctx.re2c, os.X_OK):
+        raise RuntimeError(f'Cannot find re2c executable')
+
+    test_blddir = create_tests_tree()
+    copy_tests(args.tests, test_blddir)
+
+    if os.name == 'posix':
+        subprocess.call(['chmod', '-R', 'u+w', test_blddir])
+
+    clean_test_tree(test_blddir)
+    tests = filter_tests(test_blddir)
+
+    with multiprocessing.Pool(
+            processes=jobs,
+            initializer=init_process,
+            # Note: the arguments order is important here
+            initargs=(test_blddir, args.skeleton, args.keep_temp_files,
+                      args.valgrind, args.wine,)
+    ) as pool:
+        pool.map(run_pack, chunks(tests, int(len(tests) / jobs + 1)))
```

Comment:
Shell script did a simplistic static chunking just because it's not very convenient to implement more complex pool of workers. That usually introduces bigger imbalace between chunks than single input queue thrown at a worker pool.

In python variant I suggest just using one test per work item. The following seems to speed up tests from `1.5s` to `1.2s` on 16-CPU x86_64-linux machine:

```diff
--- a/run_tests.py.in
+++ b/run_tests.py.in
@@ -600,7 +600,7 @@ def main():
             initargs=(test_blddir, args.skeleton, args.keep_temp_files,
                       args.valgrind, args.wine,)
     ) as pool:
-        pool.map(run_pack, chunks(tests, int(len(tests) / jobs + 1)))
+        pool.map(run_pack, chunks(tests, 1))

     # collect results from all threads
     total = len(tests)
```

The caveat is that stats are somehow broken (though all tests we ran for real):

```
-----------------
All:         1605
Ran:         16
Passed:      16
Soft errors: 0
Hard errors: 0
-----------------
```

Reply:
I did this initially :) But after I saw that the reports were broken, I adjusted the behavior to the bash implementation

Reply:
Let's merge the current version as is (with chunk size as in the old bash script), and then fix the stats and adjust chunk size in a follow-up change.